File: brain-move-web/src/games/memory-carte.ts
import { BaseGame } from './base-game'
import { Difficulty } from '../engine/difficulty'
import { Scoring } from '../engine/scoring'

const SYMBOLS: Record<string, string> = {
  gatto: "\u{1F431}", cane: "\u{1F436}", leone: "\u{1F981}",
  coniglio: "\u{1F430}", tartaruga: "\u{1F422}", delfino: "\u{1F42C}",
  farfalla: "\u{1F98B}", uccello: "\u{1F426}", stella: "\u2B50",
  cuore: "\u2764\uFE0F", sole: "\u{1F31E}", luna: "\u{1F319}",
  fiore: "\u{1F33C}", albero: "\u{1F333}", montagna: "\u26F0\uFE0F", onda: "\u{1F30A}",
}

export class MemoryCarte extends BaseGame {
  grid: string[] = []
  revealed: boolean[] = []
  matched: boolean[] = []
  firstCard: number | null = null
  secondCard: number | null = null
  waiting = false
  waitTimer = 0
  waitDelay = 1.0
  pairsFound = 0
  totalPairs = 0
  attempts = 0
  cols = 4
  rows = 3
  phase: "preview" | "playing" = "preview"
  showAllTimer = 0
  showAllDuration = 3.0

  constructor(difficulty: Difficulty, scoring: Scoring) {
    super(difficulty, scoring)
    this.name = "memory_carte"
    this.displayName = "Memory Carte"
  }

  start(): void {
    super.start()
    const level = this.difficulty.getLevel()
    if (level <= 2) { this.rows = 3; this.cols = 4 }
    else if (level <= 3) { this.rows = 4; this.cols = 4 }
    else { this.rows = 4; this.cols = 5 }
    const totalCells = this.rows * this.cols
    this.totalPairs = Math.floor(totalCells / 2)
    const symbols = Object.keys(SYMBOLS)
    const chosen = symbols.sort(() => Math.random() - 0.5).slice(0, this.totalPairs)
    const pairs = [...chosen, ...chosen].sort(() => Math.random() - 0.5)
    this.grid = pairs
    this.revealed = new Array(pairs.length).fill(true)
    this.matched = new Array(pairs.length).fill(false)
    this.firstCard = null
    this.secondCard = null
    this.waiting = false
    this.pairsFound = 0
    this.attempts = 0
    this.phase = "preview"
    this.showAllTimer = Date.now()
    this.feedbackMessage = "Ricorda le posizioni delle carte!"
  }

  update(_poseData?: any, dt?: number): void {
    if (!this.running) return
    if (this.phase === "preview") {
      if ((Date.now() - this.showAllTimer) / 1000 >= this.showAllDuration) {
        this.revealed = new Array(this.grid.length).fill(false)
        this.phase = "playing"
        this.feedbackMessage = "Trova le coppie! Seleziona due carte"
      }
    } else if (this.phase === "playing" && this.waiting) {
      const elapsed = (Date.now() - this.waitTimer) / 1000
      if (elapsed >= this.waitDelay) this._hidePair()
    }
  }

  selectCard(index: number): void {
    if (this.phase !== "playing" || this.waiting) return
    if (index < 0 || index >= this.grid.length || this.revealed[index] || this.matched[index]) return
    this.revealed[index] = true
    if (this.firstCard === null) {
      this.firstCard = index
      this.feedbackMessage = "Seleziona un'altra carta"
    } else if (this.secondCard === null && index !== this.firstCard) {
      this.secondCard = index
      this.attempts++
      this._checkMatch()
    }
  }

  private _checkMatch(): void {
    const i = this.firstCard!, j = this.secondCard!
    if (this.grid[i] === this.grid[j]) {
      this.matched[i] = true; this.matched[j] = true
      this.pairsFound++
      this.scoring.addScore(this.name, 15)
      this.feedbackMessage = `Coppia trovata: ${SYMBOLS[this.grid[i]]}!`
      this.firstCard = null; this.secondCard = null
      if (this.pairsFound === this.totalPairs) this._finishGame()
    } else {
      this.waiting = true
      this.waitTimer = Date.now()
      this.waitDelay = 1.0
      this.feedbackMessage = "Non corrispondono..."
    }
  }

  private _hidePair(): void {
    if (this.firstCard !== null) this.revealed[this.firstCard] = false
    if (this.secondCard !== null) this.revealed[this.secondCard] = false
    this.firstCard = null; this.secondCard = null
    this.waiting = false
    this.feedbackMessage = "Trova le coppie!"
  }

  private _finishGame(): void {
    this.stop()
    const accuracy = this.pairsFound / Math.max(this.attempts, 1)
    this.result = { completed: true, accuracy: Math.min(accuracy, 1), pairsFound: this.pairsFound, totalPairs: this.totalPairs, attempts: this.attempts }
    this.difficulty.adapt(Math.min(accuracy, 1))
    this.scoring.addScore(this.name, Math.round(accuracy * 40))
  }

  getCardSymbol(index: number): string { return SYMBOLS[this.grid[index]] ?? "?" }
  isCardRevealed(index: number): boolean { return this.revealed[index] || this.matched[index] }
  isCardMatched(index: number): boolean { return this.matched[index] }
  getGridSize(): [number, number] { return [this.rows, this.cols] }
  getProgress(): number { return this.pairsFound / Math.max(this.totalPairs, 1) }
}

File: brain-move/games/memory_carte.py
import random
import time
from games.base_game import BaseGame


class MemoryCarte(BaseGame):
    """Gioco memoria: trova le coppie di carte con simboli e animali."""

    SYMBOLS = {
        "gatto": "\U0001F431",
        "cane": "\U0001F436",
        "leone": "\U0001F981",
        "coniglio": "\U0001F430",
        "tartaruga": "\U0001F422",
        "delfino": "\U0001F42C",
        "farfalla": "\U0001F98B",
        "uccello": "\U0001F426",
        "stella": "\u2B50",
        "cuore": "\u2764\uFE0F",
        "sole": "\U0001F31E",
        "luna": "\U0001F319",
        "fiore": "\U0001F33C",
        "albero": "\U0001F333",
        "montagna": "\u26F0\uFE0F",
        "onda": "\U0001F30A",
    }

    def __init__(self, difficulty_engine, scoring_engine):
        super().__init__(difficulty_engine, scoring_engine)
        self.name = "memory_carte"
        self.display_name = "Memory Carte"
        self.grid = []
        self.revealed = []
        self.matched = []
        self.first_card = None
        self.second_card = None
        self.waiting = False
        self.wait_timer = 0
        self.wait_delay = 1.0
        self.pairs_found = 0
        self.total_pairs = 0
        self.attempts = 0
        self.max_attempts = 20
        self.cols = 4
        self.rows = 3
        self.show_all_timer = 0
        self.show_all_duration = 3.0
        self.phase = "preview"

    def start(self):
        super().start()
        level = self.difficulty.get_level()
        if level <= 2:
            self.rows, self.cols = 3, 4
        elif level <= 3:
            self.rows, self.cols = 4, 4
        else:
            self.rows, self.cols = 4, 5

        total_cells = self.rows * self.cols
        self.total_pairs = total_cells // 2
        symbols = list(self.SYMBOLS.keys())
        chosen = random.sample(symbols, self.total_pairs)
        pairs = chosen * 2
        random.shuffle(pairs)
        self.grid = pairs
        self.revealed = [False] * len(self.grid)
        self.matched = [False] * len(self.grid)
        self.first_card = None
        self.second_card = None
        self.waiting = False
        self.pairs_found = 0
        self.attempts = 0
        self.phase = "preview"
        self.show_all_timer = time.time()
        self.revealed = [True] * len(self.grid)
        self.feedback_message = "Ricorda le posizioni delle carte!"

    def update(self, pose_data=None, dt=0):
        if not self.running:
            return

        if self.phase == "preview":
            elapsed = time.time() - self.show_all_timer
            if elapsed >= self.show_all_duration:
                self.revealed = [False] * len(self.grid)
                self.phase = "playing"
                self.feedback_message = "Trova le coppie! Seleziona due carte"

        elif self.phase == "playing" and self.waiting:
            elapsed = time.time() - self.wait_timer
            if elapsed >= self.wait_delay:
                self._hide_pair()

    def select_card(self, index):
        if self.phase != "playing" or self.waiting:
            return
        if index < 0 or index >= len(self.grid):
            return
        if self.revealed[index] or self.matched[index]:
            return

        self.revealed[index] = True

        if self.first_card is None:
            self.first_card = index
            self.feedback_message = "Seleziona un'altra carta"
        elif self.second_card is None and index != self.first_card:
            self.second_card = index
            self.attempts += 1
            self._check_match()

    def _check_match(self):
        i, j = self.first_card, self.second_card
        if self.grid[i] == self.grid[j]:
            self.matched[i] = True
            self.matched[j] = True
            self.pairs_found += 1
            self.scoring.add_score(self.name, 15)
            self.feedback_message = f" Coppia trovata: {self.SYMBOLS[self.grid[i]]}!"
            self.first_card = None
            self.second_card = None
            if self.pairs_found == self.total_pairs:
                self._finish_game()
        else:
            self.waiting = True
            self.wait_timer = time.time()
            self.feedback_message = "Non corrispondono..."

    def _hide_pair(self):
        if self.first_card is not None:
            self.revealed[self.first_card] = False
        if self.second_card is not None:
            self.revealed[self.second_card] = False
        self.first_card = None
        self.second_card = None
        self.waiting = False
        self.feedback_message = "Trova le coppie!"

    def _finish_game(self):
        self.stop()
        accuracy = self.pairs_found / max(self.attempts, 1)
        self.result = {
            "completed": True,
            "accuracy": min(accuracy, 1.0),
            "pairs_found": self.pairs_found,
            "total_pairs": self.total_pairs,
            "attempts": self.attempts,
        }
        self.difficulty.adapt(min(accuracy, 1.0))
        self.scoring.add_score(self.name, int(accuracy * 40))

    def get_card_symbol(self, index):
        if 0 <= index < len(self.grid):
            return self.SYMBOLS.get(self.grid[index], "?")
        return ""

    def is_card_revealed(self, index):
        if 0 <= index < len(self.grid):
            return self.revealed[index] or self.matched[index]
        return False

    def is_card_matched(self, index):
        if 0 <= index < len(self.grid):
            return self.matched[index]
        return False

    def get_grid_size(self):
        return self.rows, self.cols

    def get_progress(self):
        return self.pairs_found / max(self.total_pairs, 1)
